File: ios/ReactionPopupStyle.swift
import ExpoModulesCore
import Foundation
import UIKit

enum EmojisPopupErrorCode: String {
  case anchorNotFound = "ANCHOR_NOT_FOUND"
  case anchorNotMeasurable = "ANCHOR_NOT_MEASURABLE"
  case emptyItems = "EMPTY_ITEMS"
  case invalidParams = "INVALID_PARAMS"
  case presentationFailed = "PRESENTATION_FAILED"
}

final class EmojisPopupError: Exception, @unchecked Sendable {
  convenience init(_ code: EmojisPopupErrorCode, description: String) {
    self.init(name: "EmojisPopupError", description: description, code: code.rawValue)
  }
}

enum ReactionPopupPlacement: String {
  case above
  case below
  case auto
}

enum ReactionPopupResolvedPlacement {
  case above
  case below
}

struct ReactionPopupItemPayload {
  let id: String
  let emoji: String
  let emojiName: String

  init(dictionary: [String: Any]) throws {
    id = try ReactionPopupParser.requiredString(dictionary["id"], key: "items[].id")
    emoji = try ReactionPopupParser.requiredString(dictionary["emoji"], key: "items[].emoji")
    emojiName = try ReactionPopupParser.requiredString(dictionary["emoji_name"], key: "items[].emoji_name")
  }
}

struct ReactionPopupHaptics {
  let onOpen: Bool
  let onPlus: Bool
  let onSelect: Bool

  init(dictionary: [String: Any]?) throws {
    onOpen = try ReactionPopupParser.bool(dictionary?["onOpen"], key: "haptics.onOpen") ?? false
    onPlus = try ReactionPopupParser.bool(dictionary?["onPlus"], key: "haptics.onPlus") ?? false
    onSelect = try ReactionPopupParser.bool(dictionary?["onSelect"], key: "haptics.onSelect") ?? false
  }
}

struct ReactionPopupStyle {
  let backgroundColor: UIColor
  let backdropColor: UIColor
  let borderColor: UIColor
  let borderRadius: CGFloat
  let borderWidth: CGFloat
  let elevation: CGFloat
  let emojiFontSize: CGFloat
  let emojiPopScale: CGFloat
  let gap: CGFloat
  let itemBackgroundColor: UIColor
  let itemBorderColor: UIColor
  let itemBorderRadius: CGFloat
  let itemBorderWidth: CGFloat
  let itemPressedBackgroundColor: UIColor
  let itemSelectedBackgroundColor: UIColor
  let itemSize: CGFloat
  let paddingHorizontal: CGFloat
  let paddingVertical: CGFloat
  let plusBackgroundColor: UIColor
  let plusIconColor: UIColor
  let plusPressedBackgroundColor: UIColor
  let shadowColor: UIColor
  let shadowOffsetX: CGFloat
  let shadowOffsetY: CGFloat
  let shadowOpacity: Float
  let shadowRadius: CGFloat
  let hoverScale: CGFloat
  let hoverTranslationY: CGFloat
  let hoverLabelBackgroundColor: UIColor
  let hoverLabelBorderRadius: CGFloat
  let hoverLabelColor: UIColor
  let hoverLabelFontSize: CGFloat
  let hoverLabelPaddingHorizontal: CGFloat
  let hoverLabelPaddingVertical: CGFloat

  init(dictionary: [String: Any]?) throws {
    let backdropBaseColor = try ReactionPopupParser.color(
      dictionary?["backdropColor"],
      key: "style.backdropColor"
    ) ?? UIColor.black
    let backdropOpacity = try ReactionPopupParser.cgFloat(
      dictionary?["backdropOpacity"],
      key: "style.backdropOpacity"
    ) ?? 0.16

    backgroundColor = try ReactionPopupParser.color(
      dictionary?["backgroundColor"],
      key: "style.backgroundColor"
    ) ?? UIColor(red: 0x1F/255.0, green: 0x1F/255.0, blue: 0x1F/255.0, alpha: 1)
    backdropColor = backdropBaseColor.withAlphaComponent(max(0, min(1, backdropOpacity)))
    borderColor = try ReactionPopupParser.color(
      dictionary?["borderColor"],
      key: "style.borderColor"
    ) ?? UIColor.clear
    borderRadius = try ReactionPopupParser.cgFloat(
      dictionary?["borderRadius"],
      key: "style.borderRadius"
    ) ?? 28
    borderWidth = try ReactionPopupParser.cgFloat(
      dictionary?["borderWidth"],
      key: "style.borderWidth"
    ) ?? 0
    elevation = try ReactionPopupParser.cgFloat(
      dictionary?["elevation"],
      key: "style.elevation"
    ) ?? 10
    emojiFontSize = try ReactionPopupParser.cgFloat(
      dictionary?["emojiFontSize"],
      key: "style.emojiFontSize"
    ) ?? 26
    emojiPopScale = try ReactionPopupParser.cgFloat(
      dictionary?["emojiPopScale"],
      key: "style.emojiPopScale"
    ) ?? 1.12
    gap = try ReactionPopupParser.cgFloat(
      dictionary?["gap"],
      key: "style.gap"
    ) ?? 10
    itemBackgroundColor = try ReactionPopupParser.color(
      dictionary?["itemBackgroundColor"],
      key: "style.itemBackgroundColor"
    ) ?? UIColor.clear
    itemBorderColor = try ReactionPopupParser.color(
      dictionary?["itemBorderColor"],
      key: "style.itemBorderColor"
    ) ?? UIColor.clear
    itemBorderRadius = try ReactionPopupParser.cgFloat(
      dictionary?["itemBorderRadius"],
      key: "style.itemBorderRadius"
    ) ?? 20
    itemBorderWidth = try ReactionPopupParser.cgFloat(
      dictionary?["itemBorderWidth"],
      key: "style.itemBorderWidth"
    ) ?? 0
    itemSelectedBackgroundColor = try ReactionPopupParser.color(
      dictionary?["itemSelectedBackgroundColor"],
      key: "style.itemSelectedBackgroundColor"
    ) ?? UIColor.tertiarySystemFill
    itemPressedBackgroundColor = try ReactionPopupParser.color(
      dictionary?["itemPressedBackgroundColor"],
      key: "style.itemPressedBackgroundColor"
    ) ?? itemSelectedBackgroundColor.withAlphaComponent(0.8)
    itemSize = try ReactionPopupParser.cgFloat(
      dictionary?["itemSize"],
      key: "style.itemSize"
    ) ?? (emojiFontSize + 16)
    paddingHorizontal = try ReactionPopupParser.cgFloat(
      dictionary?["paddingHorizontal"],
      key: "style.paddingHorizontal"
    ) ?? 12
    paddingVertical = try ReactionPopupParser.cgFloat(
      dictionary?["paddingVertical"],
      key: "style.paddingVertical"
    ) ?? 10
    plusBackgroundColor = try ReactionPopupParser.color(
      dictionary?["plusBackgroundColor"],
      key: "style.plusBackgroundColor"
    ) ?? itemBackgroundColor
    plusIconColor = try ReactionPopupParser.color(
      dictionary?["plusIconColor"],
      key: "style.plusIconColor"
    ) ?? UIColor.secondaryLabel
    plusPressedBackgroundColor = try ReactionPopupParser.color(
      dictionary?["plusPressedBackgroundColor"],
      key: "style.plusPressedBackgroundColor"
    ) ?? itemPressedBackgroundColor
    shadowColor = try ReactionPopupParser.color(
      dictionary?["shadowColor"],
      key: "style.shadowColor"
    ) ?? UIColor.black
    shadowOffsetX = try ReactionPopupParser.cgFloat(
      dictionary?["shadowOffsetX"],
      key: "style.shadowOffsetX"
    ) ?? 0
    shadowOffsetY = try ReactionPopupParser.cgFloat(
      dictionary?["shadowOffsetY"],
      key: "style.shadowOffsetY"
    ) ?? 8
    shadowOpacity = Float(
      max(
        0,
        min(
          1,
          try ReactionPopupParser.cgFloat(dictionary?["shadowOpacity"], key: "style.shadowOpacity") ?? 0.18
        )
      )
    )
    shadowRadius = try ReactionPopupParser.cgFloat(
      dictionary?["shadowRadius"],
      key: "style.shadowRadius"
    ) ?? 16
    hoverScale = try ReactionPopupParser.cgFloat(
      dictionary?["hoverScale"],
      key: "style.hoverScale"
    ) ?? 1.6
    hoverTranslationY = try ReactionPopupParser.cgFloat(
      dictionary?["hoverTranslationY"],
      key: "style.hoverTranslationY"
    ) ?? 14
    hoverLabelBackgroundColor = try ReactionPopupParser.color(
      dictionary?["hoverLabelBackgroundColor"],
      key: "style.hoverLabelBackgroundColor"
    ) ?? UIColor.black.withAlphaComponent(0.75)
    hoverLabelBorderRadius = try ReactionPopupParser.cgFloat(
      dictionary?["hoverLabelBorderRadius"],
      key: "style.hoverLabelBorderRadius"
    ) ?? 10
    hoverLabelColor = try ReactionPopupParser.color(
      dictionary?["hoverLabelColor"],
      key: "style.hoverLabelColor"
    ) ?? UIColor.white
    hoverLabelFontSize = try ReactionPopupParser.cgFloat(
      dictionary?["hoverLabelFontSize"],
      key: "style.hoverLabelFontSize"
    ) ?? 12
    hoverLabelPaddingHorizontal = try ReactionPopupParser.cgFloat(
      dictionary?["hoverLabelPaddingHorizontal"],
      key: "style.hoverLabelPaddingHorizontal"
    ) ?? 8
    hoverLabelPaddingVertical = try ReactionPopupParser.cgFloat(
      dictionary?["hoverLabelPaddingVertical"],
      key: "style.hoverLabelPaddingVertical"
    ) ?? 4
  }
}

struct ReactionPopupAnimation {
  let emojiPopScale: CGFloat
  let itemStagger: TimeInterval
  let openDuration: TimeInterval
  let trayInitialScale: CGFloat

  init(dictionary: [String: Any]?, style: ReactionPopupStyle) throws {
    emojiPopScale = try ReactionPopupParser.cgFloat(
      dictionary?["emojiPopScale"],
      key: "animation.emojiPopScale"
    ) ?? style.emojiPopScale
    itemStagger = TimeInterval(
      (try ReactionPopupParser.cgFloat(dictionary?["itemStaggerMs"], key: "animation.itemStaggerMs") ?? 18) / 1000
    )
    openDuration = TimeInterval(
      (try ReactionPopupParser.cgFloat(dictionary?["openDurationMs"], key: "animation.openDurationMs") ?? 180) / 1000
    )
    trayInitialScale = try ReactionPopupParser.cgFloat(
      dictionary?["trayInitialScale"],
      key: "animation.trayInitialScale"
    ) ?? 0.96
  }
}

struct ReactionPopupShowParams {
  let anchorId: String
  let items: [ReactionPopupItemPayload]
  let selectedId: String?
  let centerOnScreen: Bool
  let dismissOnBackdropPress: Bool
  let dismissOnDragOut: Bool
  let edgePadding: CGFloat
  let plusEnabled: Bool
  let plusAccessibilityLabel: String
  let preferredPlacement: ReactionPopupPlacement
  let showLabels: Bool
  let hideLabelsInSafeArea: Bool
  let haptics: ReactionPopupHaptics
  let animation: ReactionPopupAnimation
  let style: ReactionPopupStyle

  init(dictionary: [String: Any]) throws {
    anchorId = try ReactionPopupParser.requiredString(dictionary["anchorId"], key: "anchorId")

    let rawItems = try ReactionPopupParser.dictionaryArray(dictionary["items"], key: "items")
    guard !rawItems.isEmpty else {
      throw EmojisPopupError(
        .emptyItems,
        description: "items must contain at least one reaction"
      )
    }

    items = try rawItems.map(ReactionPopupItemPayload.init(dictionary:))
    selectedId = try ReactionPopupParser.optionalString(dictionary["selectedId"], key: "selectedId")
    centerOnScreen = try ReactionPopupParser.bool(
      dictionary["centerOnScreen"],
      key: "centerOnScreen"
    ) ?? false
    dismissOnBackdropPress = try ReactionPopupParser.bool(
      dictionary["dismissOnBackdropPress"],
      key: "dismissOnBackdropPress"
    ) ?? true
    dismissOnDragOut = try ReactionPopupParser.bool(
      dictionary["dismissOnDragOut"],
      key: "dismissOnDragOut"
    ) ?? false
    edgePadding = max(
      0,
      try ReactionPopupParser.cgFloat(dictionary["edgePadding"], key: "edgePadding") ?? 12
    )
    plusEnabled = try ReactionPopupParser.bool(dictionary["plusEnabled"], key: "plusEnabled") ?? false
    plusAccessibilityLabel = try ReactionPopupParser.optionalString(
      dictionary["plusAccessibilityLabel"],
      key: "plusAccessibilityLabel"
    ) ?? "More reactions"

    let preferredPlacementString = try ReactionPopupParser.optionalString(
      dictionary["preferredPlacement"],
      key: "preferredPlacement"
    ) ?? ReactionPopupPlacement.auto.rawValue
    guard let preferredPlacement = ReactionPopupPlacement(rawValue: preferredPlacementString) else {
      throw EmojisPopupError(
        .invalidParams,
        description: "preferredPlacement must be one of auto, above, or below"
      )
    }
    self.preferredPlacement = preferredPlacement
    showLabels = try ReactionPopupParser.bool(
      dictionary["showLabels"],
      key: "showLabels"
    ) ?? true
    hideLabelsInSafeArea = try ReactionPopupParser.bool(
      dictionary["hideLabelsInSafeArea"],
      key: "hideLabelsInSafeArea"
    ) ?? true

    style = try ReactionPopupStyle(
      dictionary: try ReactionPopupParser.optionalDictionary(dictionary["style"], key: "style")
    )
    haptics = try ReactionPopupHaptics(
      dictionary: try ReactionPopupParser.optionalDictionary(dictionary["haptics"], key: "haptics")
    )
    animation = try ReactionPopupAnimation(
      dictionary: try ReactionPopupParser.optionalDictionary(dictionary["animation"], key: "animation"),
      style: style
    )
  }
}

private enum ReactionPopupParser {
  static func requiredString(_ value: Any?, key: String) throws -> String {
    guard let string = try optionalString(value, key: key) else {
      throw EmojisPopupError(
        .invalidParams,
        description: "\(key) must be a non-empty string"
      )
    }
    return string
  }

  static func optionalString(_ value: Any?, key: String) throws -> String? {
    guard let value, let string = value as? String else {
      return nil
    }
    let trimmed = string.trimmingCharacters(in: .whitespacesAndNewlines)
    return trimmed.isEmpty ? nil : trimmed
  }

  static func bool(_ value: Any?, key: String) throws -> Bool? {
    guard let value, !(value is NSNull) else {
      return nil
    }
    if let bool = value as? Bool {
      return bool
    }
    if let number = value as? NSNumber {
      return number.boolValue
    }
    throw EmojisPopupError(
      .invalidParams,
      description: "\(key) must be a boolean"
    )
  }

  static func cgFloat(_ value: Any?, key: String) throws -> CGFloat? {
    guard let value, !(value is NSNull) else {
      return nil
    }

    let number: Double?
    if let value = value as? CGFloat {
      number = Double(value)
    } else if let value = value as? Double {
      number = value
    } else if let value = value as? Float {
      number = Double(value)
    } else if let value = value as? Int {
      number = Double(value)
    } else if let value = value as? NSNumber {
      number = value.doubleValue
    } else {
      number = nil
    }

    guard let number, number.isFinite else {
      throw EmojisPopupError(
        .invalidParams,
        description: "\(key) must be a finite number"
      )
    }
    return CGFloat(number)
  }

  static func dictionaryArray(_ value: Any?, key: String) throws -> [[String: Any]] {
    guard let value else {
      throw EmojisPopupError(
        .invalidParams,
        description: "\(key) must be an array of objects"
      )
    }
    guard let array = value as? [[String: Any]] else {
      throw EmojisPopupError(
        .invalidParams,
        description: "\(key) must be an array of objects"
      )
    }
    return array
  }

  static func optionalDictionary(_ value: Any?, key: String) throws -> [String: Any]? {
    guard let value, !(value is NSNull) else {
      return nil
    }
    guard let dictionary = value as? [String: Any] else {
      throw EmojisPopupError(
        .invalidParams,
        description: "\(key) must be an object"
      )
    }
    return dictionary
  }

  static func color(_ value: Any?, key: String) throws -> UIColor? {
    guard let value, !(value is NSNull) else {
      return nil
    }
    guard let hex = value as? String, let color = UIColor(emojisPopupHexString: hex) else {
      throw EmojisPopupError(
        .invalidParams,
        description: "\(key) must be a valid hex color string"
      )
    }
    return color
  }
}

private extension UIColor {
  convenience init?(emojisPopupHexString: String) {
    let sanitized = emojisPopupHexString
      .trimmingCharacters(in: .whitespacesAndNewlines)
      .replacingOccurrences(of: "#", with: "")

    let expanded: String
    switch sanitized.count {
    case 3:
      expanded = sanitized.map { "\($0)\($0)" }.joined() + "FF"
    case 4:
      expanded = sanitized.map { "\($0)\($0)" }.joined()
    case 6:
      expanded = sanitized + "FF"
    case 8:
      expanded = sanitized
    default:
      return nil
    }

    var rgbaValue: UInt64 = 0
    guard Scanner(string: expanded).scanHexInt64(&rgbaValue) else {
      return nil
    }

    let red = CGFloat((rgbaValue & 0xFF00_0000) >> 24) / 255
    let green = CGFloat((rgbaValue & 0x00FF_0000) >> 16) / 255
    let blue = CGFloat((rgbaValue & 0x0000_FF00) >> 8) / 255
    let alpha = CGFloat(rgbaValue & 0x0000_00FF) / 255

    self.init(red: red, green: green, blue: blue, alpha: alpha)
  }
}

File: android/src/main/java/expo/modules/nativeemojispopup/ReactionPopupStyle.kt
package expo.modules.nativeemojispopup

import android.graphics.Color
import expo.modules.kotlin.exception.CodedException

enum class EmojisPopupErrorCode(val value: String) {
  ANCHOR_NOT_FOUND("ANCHOR_NOT_FOUND"),
  ANCHOR_NOT_MEASURABLE("ANCHOR_NOT_MEASURABLE"),
  EMPTY_ITEMS("EMPTY_ITEMS"),
  INVALID_PARAMS("INVALID_PARAMS"),
  PRESENTATION_FAILED("PRESENTATION_FAILED"),
}

class EmojisPopupException(
  errorCode: EmojisPopupErrorCode,
  message: String,
) : CodedException(errorCode.value, message, null)

enum class ReactionPopupPlacement {
  ABOVE,
  BELOW,
  AUTO,
  ;

  companion object {
    fun from(value: String?): ReactionPopupPlacement =
      when (value) {
        "above" -> ABOVE
        "below" -> BELOW
        else -> AUTO
      }
  }
}

data class ReactionPopupItemPayload(
  val id: String,
  val emoji: String,
  val emojiName: String,
) {
  companion object {
    fun fromMap(map: Map<String, Any?>): ReactionPopupItemPayload {
      val id = (map["id"] as? String)?.trim().orEmpty()
      val emoji = map["emoji"] as? String ?: ""
      val emojiName = (map["emoji_name"] as? String)?.trim().orEmpty()

      if (id.isEmpty() || emoji.isEmpty() || emojiName.isEmpty()) {
        throw EmojisPopupException(
          EmojisPopupErrorCode.INVALID_PARAMS,
          "Emojis popup items require non-empty id, emoji, and emoji_name.",
        )
      }

      return ReactionPopupItemPayload(
        id = id,
        emoji = emoji,
        emojiName = emojiName,
      )
    }
  }
}

data class ReactionPopupHaptics(
  val onOpen: Boolean = false,
  val onPlus: Boolean = false,
  val onSelect: Boolean = false,
) {
  companion object {
    fun fromMap(map: Map<String, Any?>?): ReactionPopupHaptics =
      ReactionPopupHaptics(
        onOpen = map?.get("onOpen") as? Boolean ?: false,
        onPlus = map?.get("onPlus") as? Boolean ?: false,
        onSelect = map?.get("onSelect") as? Boolean ?: false,
      )
  }
}

data class ReactionPopupAnimation(
  val emojiPopScale: Float = 1.12f,
  val itemStaggerMs: Long = 18L,
  val openDurationMs: Long = 180L,
  val trayInitialScale: Float = 0.96f,
) {
  companion object {
    fun fromMap(map: Map<String, Any?>?): ReactionPopupAnimation =
      ReactionPopupAnimation(
        emojiPopScale = (map?.get("emojiPopScale") as? Number)?.toFloat() ?: 1.12f,
        itemStaggerMs = (map?.get("itemStaggerMs") as? Number)?.toLong() ?: 18L,
        openDurationMs = (map?.get("openDurationMs") as? Number)?.toLong() ?: 180L,
        trayInitialScale = (map?.get("trayInitialScale") as? Number)?.toFloat() ?: 0.96f,
      )
  }
}

data class ReactionPopupStyle(
  val backgroundColor: Int = colorOrNull(null) ?: Color.parseColor("#1F1F1F"),
  val backdropColor: Int = colorOrNull(null) ?: Color.BLACK,
  val backdropOpacity: Float = 0.16f,
  val borderColor: Int = Color.TRANSPARENT,
  val borderRadius: Float = 28f,
  val borderWidth: Float = 0f,
  val elevation: Float = 10f,
  val emojiFontSize: Float = 26f,
  val gap: Float = 10f,
  val itemBackgroundColor: Int = Color.TRANSPARENT,
  val itemBorderColor: Int = Color.TRANSPARENT,
  val itemBorderRadius: Float = 20f,
  val itemBorderWidth: Float = 0f,
  val itemPressedBackgroundColor: Int = Color.argb(31, 255, 255, 255),
  val itemSelectedBackgroundColor: Int = Color.argb(36, 255, 255, 255),
  val itemSize: Float = 40f,
  val paddingHorizontal: Float = 12f,
  val paddingVertical: Float = 10f,
  val plusBackgroundColor: Int = Color.argb(20, 255, 255, 255),
  val plusIconColor: Int = Color.parseColor("#AFAFAF"),
  val plusPressedBackgroundColor: Int = Color.argb(41, 255, 255, 255),
  val shadowColor: Int = Color.BLACK,
  val hoverScale: Float = 1.6f,
  val hoverTranslationY: Float = 14f,
  val hoverLabelBackgroundColor: Int = Color.argb(192, 0, 0, 0),
  val hoverLabelBorderRadius: Float = 10f,
  val hoverLabelColor: Int = Color.WHITE,
  val hoverLabelFontSize: Float = 12f,
  val hoverLabelPaddingHorizontal: Float = 8f,
  val hoverLabelPaddingVertical: Float = 4f,
) {
  companion object {
    fun fromMap(map: Map<String, Any?>?): ReactionPopupStyle {
      val parsedEmojiFontSize = (map?.get("emojiFontSize") as? Number)?.toFloat() ?: 26f
      return ReactionPopupStyle(
        backgroundColor = colorOrNull(map?.get("backgroundColor") as? String)
          ?: Color.parseColor("#1F1F1F"),
        backdropColor = colorOrNull(map?.get("backdropColor") as? String)
          ?: Color.BLACK,
        backdropOpacity = (map?.get("backdropOpacity") as? Number)?.toFloat() ?: 0.16f,
        borderColor = colorOrNull(map?.get("borderColor") as? String) ?: Color.TRANSPARENT,
        borderRadius = (map?.get("borderRadius") as? Number)?.toFloat() ?: 28f,
        borderWidth = (map?.get("borderWidth") as? Number)?.toFloat() ?: 0f,
        elevation = (map?.get("elevation") as? Number)?.toFloat() ?: 10f,
        emojiFontSize = parsedEmojiFontSize,
        gap = (map?.get("gap") as? Number)?.toFloat() ?: 10f,
        itemBackgroundColor = colorOrNull(map?.get("itemBackgroundColor") as? String)
          ?: Color.TRANSPARENT,
        itemBorderColor = colorOrNull(map?.get("itemBorderColor") as? String)
          ?: Color.TRANSPARENT,
        itemBorderRadius = (map?.get("itemBorderRadius") as? Number)?.toFloat() ?: 20f,
        itemBorderWidth = (map?.get("itemBorderWidth") as? Number)?.toFloat() ?: 0f,
        itemPressedBackgroundColor = colorOrNull(map?.get("itemPressedBackgroundColor") as? String)
          ?: Color.argb(31, 255, 255, 255),
        itemSelectedBackgroundColor = colorOrNull(map?.get("itemSelectedBackgroundColor") as? String)
          ?: Color.argb(36, 255, 255, 255),
        itemSize = (map?.get("itemSize") as? Number)?.toFloat() ?: (parsedEmojiFontSize + 16f),
        paddingHorizontal = (map?.get("paddingHorizontal") as? Number)?.toFloat() ?: 12f,
        paddingVertical = (map?.get("paddingVertical") as? Number)?.toFloat() ?: 10f,
        plusBackgroundColor = colorOrNull(map?.get("plusBackgroundColor") as? String)
          ?: Color.argb(20, 255, 255, 255),
        plusIconColor = colorOrNull(map?.get("plusIconColor") as? String)
          ?: Color.parseColor("#AFAFAF"),
        plusPressedBackgroundColor = colorOrNull(map?.get("plusPressedBackgroundColor") as? String)
          ?: Color.argb(41, 255, 255, 255),
        shadowColor = colorOrNull(map?.get("shadowColor") as? String) ?: Color.BLACK,
        hoverScale = (map?.get("hoverScale") as? Number)?.toFloat() ?: 1.6f,
        hoverTranslationY = (map?.get("hoverTranslationY") as? Number)?.toFloat() ?: 14f,
        hoverLabelBackgroundColor = colorOrNull(map?.get("hoverLabelBackgroundColor") as? String)
          ?: Color.argb(192, 0, 0, 0),
        hoverLabelBorderRadius = (map?.get("hoverLabelBorderRadius") as? Number)?.toFloat() ?: 10f,
        hoverLabelColor = colorOrNull(map?.get("hoverLabelColor") as? String) ?: Color.WHITE,
        hoverLabelFontSize = (map?.get("hoverLabelFontSize") as? Number)?.toFloat() ?: 12f,
        hoverLabelPaddingHorizontal = (map?.get("hoverLabelPaddingHorizontal") as? Number)?.toFloat() ?: 8f,
        hoverLabelPaddingVertical = (map?.get("hoverLabelPaddingVertical") as? Number)?.toFloat() ?: 4f,
      )
    }
  }
}

data class ReactionPopupShowParams(
  val anchorId: String,
  val items: List<ReactionPopupItemPayload>,
  val selectedId: String?,
  val centerOnScreen: Boolean,
  val dismissOnBackdropPress: Boolean,
  val dismissOnDragOut: Boolean,
  val edgePadding: Float,
  val plusEnabled: Boolean,
  val plusAccessibilityLabel: String,
  val preferredPlacement: ReactionPopupPlacement,
  val showLabels: Boolean,
  val hideLabelsInSafeArea: Boolean,
  val haptics: ReactionPopupHaptics,
  val animation: ReactionPopupAnimation,
  val style: ReactionPopupStyle,
) {
  companion object {
    fun fromMap(map: Map<String, Any?>): ReactionPopupShowParams {
      val anchorId = (map["anchorId"] as? String)?.trim().orEmpty()
      if (anchorId.isEmpty()) {
        throw EmojisPopupException(
          EmojisPopupErrorCode.INVALID_PARAMS,
          "Emojis popup requires a non-empty anchorId.",
        )
      }

      val rawItems = map["items"] as? List<Map<String, Any?>>
      if (rawItems.isNullOrEmpty()) {
        throw EmojisPopupException(
          EmojisPopupErrorCode.EMPTY_ITEMS,
          "Emojis popup requires at least one item.",
        )
      }

      return ReactionPopupShowParams(
        anchorId = anchorId,
        items = rawItems.map(ReactionPopupItemPayload::fromMap),
        selectedId = map["selectedId"] as? String,
        centerOnScreen = map["centerOnScreen"] as? Boolean ?: false,
        dismissOnBackdropPress = map["dismissOnBackdropPress"] as? Boolean ?: true,
        dismissOnDragOut = map["dismissOnDragOut"] as? Boolean ?: false,
        edgePadding = (map["edgePadding"] as? Number)?.toFloat() ?: 12f,
        plusEnabled = map["plusEnabled"] as? Boolean ?: false,
        plusAccessibilityLabel = ((map["plusAccessibilityLabel"] as? String)?.trim())
          ?.takeIf { it.isNotEmpty() }
          ?: "More reactions",
        preferredPlacement = ReactionPopupPlacement.from(map["preferredPlacement"] as? String),
        showLabels = map["showLabels"] as? Boolean ?: true,
        hideLabelsInSafeArea = map["hideLabelsInSafeArea"] as? Boolean ?: true,
        haptics = ReactionPopupHaptics.fromMap(map["haptics"] as? Map<String, Any?>),
        animation = ReactionPopupAnimation.fromMap(map["animation"] as? Map<String, Any?>),
        style = ReactionPopupStyle.fromMap(map["style"] as? Map<String, Any?>),
      )
    }
  }
}

private fun colorOrNull(value: String?): Int? {
  val normalized = value?.trim()?.takeIf { it.isNotEmpty() } ?: return null
  val hex = normalized.removePrefix("#")
  // All inputs treated as RRGGBB or RRGGBBAA (matching iOS convention).
  // Color.parseColor expects #AARRGGBB, so we reorder when alpha is present.
  val expanded = when (hex.length) {
    3 -> "#FF${hex[0]}${hex[0]}${hex[1]}${hex[1]}${hex[2]}${hex[2]}"
    4 -> "#${hex[3]}${hex[3]}${hex[0]}${hex[0]}${hex[1]}${hex[1]}${hex[2]}${hex[2]}"
    6 -> "#FF$hex"
    8 -> "#${hex.substring(6, 8)}${hex.substring(0, 6)}"
    else -> return null
  }
  return runCatching { Color.parseColor(expanded) }.getOrNull()
}
